### ISSUE TITLE
HA Frontend keeps failing over

### ISSUE THREAD

**Issue by MedicMomcilo** [OpenNebula/one#1673]
# Bug Report
## Version of OpenNebula

- [ ] 5.2.2
- [ ] 5.4.0
- [ ] 5.4.1
- [ ] 5.4.2
- [ ] 5.4.3
- [ ] 5.4.4
- [X] 5.4.5
- [ ] Development build

## Component

- [ ] Authorization (LDAP, x509 certs...)
- [ ] Command Line Interface (CLI)
- [ ] Contextualization
- [ ] Documentation
- [X] Federation and HA
- [ ] Host, Clusters and Monitorization
- [ ] KVM
- [ ] Networking
- [ ] Orchestration (OpenNebula Flow)
- [ ] Packages
- [ ] Scheduler
- [ ] Storage & Images
- [ ] Sunstone
- [ ] Upgrades
- [ ] User, Groups, VDCs and ACL
- [ ] vCenter

## Description
Cluster of three nodes was set up using official documentation[1].
However, even when upping heartbeat threshold to 25 seconds it keeps failing over.
Log contains just a note that heartbeat wasn't received and that election will take place:

> Mon Jan 22 17:37:34 2018 [Z0][ReM][D]: Req:2688 UID:0 one.zone.raftstatus result SUCCESS, "<RAFT><SERVER_ID>3</..."
> Mon Jan 22 17:38:04 2018 [Z0][ReM][D]: Req:9488 UID:0 one.zone.raftstatus invoked 
> Mon Jan 22 17:38:04 2018 [Z0][ReM][D]: Req:9488 UID:0 one.zone.raftstatus result SUCCESS, "<RAFT><SERVER_ID>3</..."
> Mon Jan 22 17:38:34 2018 [Z0][ReM][D]: Req:3904 UID:0 one.zone.raftstatus invoked 
> Mon Jan 22 17:38:34 2018 [Z0][ReM][D]: Req:3904 UID:0 one.zone.raftstatus result SUCCESS, "<RAFT><SERVER_ID>3</..."
> Mon Jan 22 17:39:04 2018 [Z0][ReM][D]: Req:9536 UID:0 one.zone.raftstatus invoked 
> **Mon Jan 22 17:39:14 2018 [Z0][RRM][E]: Failed to get heartbeat from leader. Starting election proccess**
> Mon Jan 22 17:39:15 2018 [Z0][ReM][I]: Leader term (3617) is outdated (3618)
> Mon Jan 22 17:39:16 2018 [Z0][ReM][D]: Req:9536 UID:0 one.zone.raftstatus result SUCCESS, "<RAFT><SERVER_ID>3</..."
> Mon Jan 22 17:39:16 2018 [Z0][RCM][I]: Vote not granted from follower 2: [one.zone.voterequest] Candidate's log is outdated
> Mon Jan 22 17:39:16 2018 [Z0][RCM][I]: Vote not granted from follower 4: [one.zone.voterequest] Candidate's log is outdated
> Mon Jan 22 17:39:41 2018 [Z0][ReM][D]: Req:7200 UID:0 one.zone.voterequest invoked , 3619, 4, 618248, 3617
> Mon Jan 22 17:39:41 2018 [Z0][ReM][I]: New term (3619) discovered from candidate 4
> Mon Jan 22 17:39:41 2018 [Z0][ReM][D]: Req:5360 UID:0 one.zone.voterequest invoked , 3619, 2, 618248, 3617
> Mon Jan 22 17:39:41 2018 [Z0][RCM][I]: oned is set to follower mode
> Mon Jan 22 17:39:41 2018 [Z0][ReM][I]: New term (3619) discovered from candidate 2
> Mon Jan 22 17:39:41 2018 [Z0][ReM][D]: Req:7200 UID:0 one.zone.voterequest result SUCCESS, 3618
> Mon Jan 22 17:39:41 2018 [Z0][RCM][I]: oned is set to follower mode
> Mon Jan 22 17:39:41 2018 [Z0][ReM][E]: Req:5360 UID:0 one.zone.voterequest result FAILURE [one.zone.voterequest] Already voted for another candidate
> Mon Jan 22 17:39:46 2018 [Z0][ReM][D]: Req:4528 UID:0 one.zone.raftstatus invoked 
> Mon Jan 22 17:39:46 2018 [Z0][ReM][D]: Req:4528 UID:0 one.zone.raftstatus result SUCCESS, "<RAFT><SERVER_ID>3</..."
> Mon Jan 22 17:39:58 2018 [Z0][MKP][I]: --Mark--
> Mon Jan 22 17:39:58 2018 [Z0][ImM][I]: --Mark--
> Mon Jan 22 17:39:58 2018 [Z0][InM][I]: --Mark--
> Mon Jan 22 17:39:58 2018 [Z0][VMM][I]: --Mark--
> Mon Jan 22 17:40:16 2018 [Z0][ReM][D]: Req:3936 UID:0 one.zone.raftstatus invoked 
> Mon Jan 22 17:40:16 2018 [Z0][ReM][D]: Req:3936 UID:0 one.zone.raftstatus result SUCCESS, "<RAFT><SERVER_ID>3</..."
> 

Network dump has been done at the same time with tcpdump and there were no silence at that moment.
As you can see it only passed 1m 10s since last SUCCESS, with threshold of 25 seconds that shouldn't happen.
Relevant config file part:

> 
> FEDERATION = [
>     MODE          = "STANDALONE",
>     ZONE_ID       = 0,
>     SERVER_ID     = 3,
>     MASTER_ONED   = ""
> ]
> 
> RAFT = [
>     LOG_RETENTION        = 500000,
>     LOG_PURGE_TIMEOUT    = 600,
>     ELECTION_TIMEOUT_MS  = 25000,
>     BROADCAST_TIMEOUT_MS = 500,
>     XMLRPC_TIMEOUT_MS    = 0
> ]
> 

Please let me know if there is any other relevant information I should share.

[1] http://docs.opennebula.org/5.4/advanced_components/ha/frontend_ha_setup.html

### Expected Behavior
Failover doesn't happen unless there is node failure.

### Actual Behavior
Failover happens more than once a day.

## How to reproduce
Configure OpenNebula frontend in HA mode.

**Comment by rsmontero** [comment]
Trying to understand this, the problem is that for period shorter than the threshold you are still seeing election timeouts, i.e. this bug is for oned not observing the value of ELECTION_TIMEOUT?

**Comment by MedicMomcilo** [comment]
No, ELECTION_TIMEOUT is observed and respected by oned.
I've tested this by stopping service manually and timing fail over.
Then, I increased timeout to 45 seconds and stopped service again with the same results.

So, it is not that timeout value is not taken into account, but rather it fails over without any obvious reason.

**Comment by hydro-b** [comment]
On Tue, Jan 23, 2018 at 01:42:28PM +0000, MedicMomcilo wrote:
> No, ELECTION_TIMEOUT is observed and respected by oned.
> I've tested this by stopping service manually and timing fail over.
> Then, I increased timeout to 45 seconds and stopped service again with the same results.
> 
> So, it is not that timeout value is not taken into account, but rather
> it fails over without any obvious reason.

I can confirm this behaviour. We have 3 clusters with 3 ONE nodes each
(RAFT). We see a lot of fail overs happening without obivous reasons,
and a lot of them too. No logging (we have debug turned on) as _WHY_
this happened.

Gr. Stefan

**Comment by MedicMomcilo** [comment]
Upgraded all OpenNebula components to 5.4.6, still seeing the same problem.
However, setting high ELECTION_TIMEOUT (45 seconds) does improve situation.
We're now faced with failover once every few days as opposed to few times a day.

**Comment by MedicMomcilo** [comment]
We're still facing this issue.
It is weird that there was successful check of raftstatus, and then one second later, we're getting missing heartbeat message:
```
...
Tue Apr  3 19:58:16 2018 [Z0][ReM][D]: Req:1360 UID:0 one.zone.raftstatus invoked 
Tue Apr  3 19:58:16 2018 [Z0][ReM][D]: Req:1360 UID:0 one.zone.raftstatus result SUCCESS, "<RAFT><SERVER_ID>2</..."
Tue Apr  3 19:58:17 2018 [Z0][RRM][E]: Failed to get heartbeat from leader. Starting election proccess
...
```

**Comment by atodorov-storpool** [comment]
Do you observe processes in D state when it happens?

There is a chance to catch this if you collect the processes info in the raft hook...

**Comment by rsmontero** [comment]
This is fixed in recent versions (>5.8.4) of OpenNebula

**Comment by bogdanparvu18** [comment]
Thanks Ruben.

Bogdan Parvu
